### PR TITLE
support of singleton and lock for win32 and c++ 11

### PR DESCRIPTION
**Pull request by gvollant**
The pull request #2371 added support for lock with c++ 11, but not Singleton.

This PR provide support of both singleton and lock for both C++11 and Windows native Win32 API (like posix does).

### CODE REVIEW

**Comment by gvollant** [comment]
with @googlebot I signed it!

**Comment by gvollant** [comment]
I do a rebase on 8.10.23 master

**Comment by gvollant** [comment]
> 
> 
> I do a rebase on 8.10.23 master

Sorry  dariuske , I just discovered my rebase erased your modification. Do you have a backup?

**Comment by mhmd-khalil49** [comment]
Hi Team,
Just to let you know that you should update the GABON (GA, 241) country phone number. The numbering changed last week on 15th November 2019 night.

Metadata for Gabon (GA, +241) need to be updated. They added a digit, increasing the length from 8 to 9 digits.

**Comment by penmetsaa** [comment]
Hi,
Sorry for the delay happened. Respective fix will be part of this release, that is planned to complete by end of this week. To report any such issues, please follow instructions at https://github.com/google/libphonenumber/blob/master/CONTRIBUTING.md. Pinging in the other thread may not be efficient in tracking/updating the issue later on. Thanks!

**Comment by gvollant** [comment]
Thank you, and about the original subject of this thread (the PR), what is the procedure now?

 

Regards

Gilles Vollant

 

De : penmetsaa [mailto:notifications@github.com] 
Envoyé : mardi 26 novembre 2019 06:14
À : google/libphonenumber
Cc : Gilles Vollant; Author
Objet : Re: [google/libphonenumber] support of singleton and lock for win32 and c++ 11 (#2400)

 

Hi,
Sorry for the delay happened. Respective fix will be part of this release, that is planned to complete by end of this week. To report any such issues, please follow instructions at https://github.com/google/libphonenumber/blob/master/CONTRIBUTING.md. Pinging in the other thread may not be efficient in tracking/updating the issue later on. Thanks!

**Comment by gvollant** [comment]
De : Gilles Vollant [mailto:vollant.g@gmail.com] 
Envoyé : mardi 10 décembre 2019 09:36
À : 'libphonenumber-discuss@googlegroups.com'
Objet : Again : compatibility multithread without Posix or Boost

Hello,

I wrote again to explain better my PR #2400. I really hope project maintainer will review it.

Currently, C++ LibPhoneNumber support multithread only with Posix or boost.

For Posix, the ifdef test if __APPLE__ or __linux__ are defined. But somes other Posix environnement are compatible (like IBM AIX, Android NDK…, xxxBSDs)

Boost is a very huge library and adding Boost only for singleton and mutex is not the better thing.

There was a first PR (#2371) already merged, which uses std::mutex on C++ 2011 compiler, but did not offer solution for singleton
https://github.com/google/libphonenumber/pull/2371 

And Visual Studio 2019 define __cplusplus to 199711L
https://docs.microsoft.com/en-us/cpp/build/reference/zc-cplusplus?view=vs-2019 

So my PR had singleton and mutex implementation:
-	Current Posix implementation, with detection of additional PHONENUMBERS_HAVE_POSIX_THREAD symbol
-	Current Boost implementation
-	C++2011 using std::mutex for an implementation for singleton
-	Win32 native API

Now I wait review

**Comment by gvollant** [comment]
I rebased again

**Comment by gvollant** [comment]
@dariuske are you here for review?

**Comment by gvollant** [comment]
> @dariuske are you here for review?

or @dvadym or @AlipmanG

**Comment by battre** [comment]
Hi @gvollant,
I am not a project owner but work on Chromium and rely on libphonenumber (I worked with @dvadym) on the project. I have taken your patch and cleaned it up a little bit (following naming conventions, having a consistent file name structure, updating CMakeLists.txt). The big caveat is that due to Covid 19 I did this in a Windows VM and later learned that the compiler is broken. So I could not verify whether any of the code compiles.
Could you please take a look, see whether everything compiles and maybe apply it to your pull request? I'll then try to ping one of the library owners.
Best regards,
Dominic

[patch.txt](https://github.com/google/libphonenumber/files/4706011/patch.txt)

**Comment by battre** [comment]
Ok, got the compiler to work. Here is a fixed version that I tested successfully for the std::mutex, win32 and unsafe version with VC 2019.

[patch.txt](https://github.com/google/libphonenumber/files/4706234/patch.txt)

**Comment by battre** [comment]
And posix complies happily on Linux.

**Comment by battre** [comment]
The patch should apply cleanly on revision 5e9507a46051405120bc73fcc13d0b0be1b93c29 with `patch -1 < patch.txt`.

**Comment by gvollant** [comment]
I used your patch and done three additional modification:
- when test defined(__linux__) || defined(__APPLE__), add support of PHONENUMBERS_HAVE_POSIX_THREAD, so FreeBsd/AIX and other can manually define this symbols if needed
- there was mismatch WIN32 / _WIN32 symbol test. I test always WIN32 (without _ prefix)
- I add strange compile error on cpp/src/phonenumbers/asyoutypeformatter.cc on std::min symbol (is seem protobuf define then undefing it, and including windows.h cause strange conflict). I just removed std::min and replace by simple test.

**Comment by gvollant** [comment]
> 
> 
> Hi @gvollant,
> I am not a project owner but work on Chromium and rely on libphonenumber (I worked with @dvadym) on the project. I have taken your patch and cleaned it up a little bit (following naming conventions, having a consistent file name structure, updating CMakeLists.txt). The big caveat is that due to Covid 19 I did this in a Windows VM and later learned that the compiler is broken. So I could not verify whether any of the code compiles.
> Could you please take a look, see whether everything compiles and maybe apply it to your pull request? I'll then try to ping one of the library owners.
> Best regards,
> Dominic
> 
> [patch.txt](https://github.com/google/libphonenumber/files/4706011/patch.txt)

@battre now we can probably ask a review

File: cpp/src/phonenumbers/asyoutypeformatter.cc [hunk 1]
```diff
@@ -211,9 +211,11 @@ void AsYouTypeFormatter::NarrowDownPossibleFormats(
...
       continue;
     }
-    int last_leading_digits_pattern =
-        std::min(index_of_leading_digits_pattern,
-                 format.leading_digits_pattern_size() - 1);
+    // We don't use std::min because there is stange symbol conflict
+    // with including <windows.h> and protobuf symbols
+    int last_leading_digits_pattern = format.leading_digits_pattern_size() - 1;
+    if (last_leading_digits_pattern > index_of_leading_digits_pattern)
+        last_leading_digits_pattern = index_of_leading_digits_pattern;
```
Comment (battre) [inline cpp/src/phonenumbers/asyoutypeformatter.cc:218]: -2 spaces indentation

File: cpp/src/phonenumbers/base/memory/singleton_boost.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
+// Copyright (C) 2013 The Libphonenumber Authors
```
Comment (battre) [inline cpp/src/phonenumbers/base/memory/singleton_boost.h:1]: given that this is a new file, we should probably update this to 2020

File: cpp/src/phonenumbers/base/thread_checker.h [hunk 3]
```diff
@@ -22,16 +22,18 @@
 // Note that I18N_PHONENUMBERS_NO_THREAD_SAFETY must be defined only to let the
 // user of the library know that it can't be used in a thread-safe manner when
 // it is not depending on Boost.
-#if !defined(__linux__) && !defined(__APPLE__) && \
-    !defined(I18N_PHONENUMBERS_NO_THREAD_SAFETY)
+#if !defined(__linux__) && !defined(__APPLE__) && !defined(PHONENUMBERS_HAVE_POSIX_THREAD) && \
+    !defined(I18N_PHONENUMBERS_NO_THREAD_SAFETY) && \
+	!((__cplusplus>=201103L) && (!(defined(NO_LIBPHONENUMBERS_STDMUTEX)))) && \
```
Comment (battre) [inline cpp/src/phonenumbers/base/thread_checker.h:27]: nit: spaces around >=
nit: please replace (!( with !

If you want to support NO_LIBPHONENUMBERS_STDMUTEX, please add it to lock.h, singleton.h and CMakeLists.txt.

File: cpp/src/phonenumbers/base/memory/singleton.h [hunk 4]
```diff
@@ -18,78 +18,16 @@
...
-
-}  // namespace phonenumbers
-}  // namespace i18n
-
-#else
+#include "phonenumbers/base/memory/singleton_boost.h"
+#elif __cplusplus >= 201103L
+// C++11 Lock implementation based on std::mutex.
+#include "phonenumbers/base/memory/singleton_stdmutex.h"
+#elif defined(__linux__) || defined(__APPLE__) || defined(PHONENUMBERS_HAVE_POSIX_THREAD)
```
Comment (battre) [inline cpp/src/phonenumbers/base/memory/singleton.h:25]: Please rename PHONENUMBERS_HAVE_POXIX_THREAD to I18N_... (also in other places) to be in line with the I18N_PHONENUMBERS_USE_BOOST naming convention.

Please add it to CMakeLists.txt.

Comment (gvollant) [reply cpp/src/phonenumbers/base/memory/singleton.h]: I made the modification and add text to readme

You can now commit on my fork repo
https://github.com/gvollant/libphonenumber , branch gvollant/lock_singleton_win32_cpp11

Commiting in this branch is the better solution

File: cpp/CMakeLists.txt [hunk 5]
```diff
@@ -104,6 +104,14 @@ if (${USE_BOOST} STREQUAL "ON")
   include_directories (${Boost_INCLUDE_DIRS})
 endif ()
 
+if (${NOT_USE_STDMUTEX} STREQUAL "ON")
```
Comment (battre) [inline cpp/CMakeLists.txt:107]: Can you add these to the options above (line 80 and following)?

Comment (gvollant) [reply cpp/CMakeLists.txt:109]: done

File: cpp/src/phonenumbers/base/memory/singleton_boost.h (hunk 2 above)
Comment (gvollant) [reply cpp/src/phonenumbers/base/memory/singleton_boost.h]: done (but code itself is older)

**Comment by gvollant** [comment]
> One final request. :-)

@battre is it good now?
I compiled with VS2019 under windows and cmake + clang under linux

**Comment by gvollant** [comment]
> Here lock_boost.h comes with its own AutoLock. Could you please wrap the AutLock implementation in lock.h in
> #if !defined(I18N_PHONENUMBERS_USE_BOOST)
> #endif
> ?

done

**Comment by gvollant** [comment]
> 
> 
> Thanks a lot!
thank you a lot also !
@penmetsaa what is the next step?

**Comment by battre** [comment]
In the past I only tried to compile this without cmake. I found that on Linux it did not link.

Here is another patch. With that I could ensure that everything works with boost, std::mutex and pthread. I could not try it out with -DBUILD_STATIC_LIB=OFF because that seems to fail regardless of the patch. It looks like the CMakeLists.txt file did not get a lot of love recently.  Unfortunately, I am not an expert either.

[cmake-fix.txt](https://github.com/google/libphonenumber/files/4719724/cmake-fix.txt)

**Comment by gvollant** [comment]
> 
> 
> In the past I only tried to compile this without cmake. I found that on Linux it did not link.
> 
> Here is another patch. With that I could ensure that everything works with boost, std::mutex and pthread. I could not try it out with -DBUILD_STATIC_LIB=OFF because that seems to fail regardless of the patch. It looks like the CMakeLists.txt file did not get a lot of love recently. Unfortunately, I am not an expert either.
> 
> [cmake-fix.txt](https://github.com/google/libphonenumber/files/4719724/cmake-fix.txt)

instead patch, can you try commit at my fork https://github.com/gvollant/libphonenumber , branch gvollant/lock_singleton_win32_cpp11 ?

**Comment by gvollant** [comment]
> 
> 
> In the past I only tried to compile this without cmake. I found that on Linux it did not link.
> 
> Here is another patch. With that I could ensure that everything works with boost, std::mutex and pthread. I could not try it out with -DBUILD_STATIC_LIB=OFF because that seems to fail regardless of the patch. It looks like the CMakeLists.txt file did not get a lot of love recently. Unfortunately, I am not an expert either.
> 
> [cmake-fix.txt](https://github.com/google/libphonenumber/files/4719724/cmake-fix.txt)

I applied your patch
You need review again, the reviewed statut has gone with modification

File: cpp/CMakeLists.txt [hunk 6]
```diff
@@ -415,6 +430,8 @@ set (COMMON_DEPS ${ICU_LIB})
 
 set (LIBRARY_DEPS ${PROTOBUF_LIB})
 
+set (COMPILE_OPTIONS ())
```
Comment (battre) [inline cpp/CMakeLists.txt:433]: sorry, this should go away.

Comment (gvollant) [reply cpp/CMakeLists.txt]: done @battre

**Comment by gvollant** [comment]
> 
> 
> I have verified that this compiles for Chrome on Windows and stand-alone on Linux (in the variations boost, std::mutex and pthreads). I think we are good to go now. If there is more work on the CMake configuration needed, I think that people who rely on it will speak up. Given that `cmake -DBUILD_STATIC_LIB=OFF` is broken, I guess that most users rely on their on build files.

I made a small other PR #2475 to compiler libphonernumber without java (susing a first work from @Venemo )

**Comment by penmetsaa** [comment]
Thanks for your patience Gilles. With your consent, we would like to add you to our contributors list: https://github.com/google/libphonenumber/pull/2479

**Comment by penmetsaa** [comment]
Gilles, could you please update the branch?
so that I can proceed with the final merge.

**Comment by gvollant** [comment]
@battre the commit is merged, but on a test with USE_POSIX_THREAD=ON

I've the error
clang: error: no input files
/bin/sh: 1: -pthread: not found


are you sure that, we added
list (APPEND CMAKE_C_FLAGS "-pthread")

pehaps we need 
list (APPEND CMAKE_C_FLAGS "-lpthread")

replace -pthread by -lpthread

**Comment by gvollant** [comment]
I've error

[ 65%] Building CXX object CMakeFiles/phonenumber.dir/src/phonenumbers/utf/unicodetext.cc.o
gcc: fatal error: no input files
compilation terminated.
/bin/sh: 1: -pthread: not found
make[3]: *** [CMakeFiles/phonenumber.dir/build.make:273: CMakeFiles/phonenumber.dir/src/phonenumbers/utf/rune.c.o] Error 127
make[3]: *** Waiting for unfinished jobs....
[ 69%] Building CXX object CMakeFiles/phonenumber.dir/src/phonenumbers/utf/unilib.cc.o

**Comment by gvollant** [comment]
@battre ?

**Comment by battre** [comment]
Have you tried whether it works with `-lpthread`? :-)

It looks to me like `/bin/sh` was trying to execute `-pthread`. The fact that gcc sees no input files is also strange.

What do you get if you execute `make SHELL="/bin/bash -x"` ?

**Comment by battre** [comment]
@gvollant ?

**Comment by gvollant** [comment]
I will test.
Note @battre : can you email me at info AT winimage.com ?

**Comment by gvollant** [comment]
I have the problem when I use CFLAGS when compile utf\rune.c

example:
export CFLAGS=" -fdata-sections  "
export CFLAGS=" -fPIC "
before CMAKE

no problem with CXXFLAGS

with previuos version , I used both CSSFLAGS and CFLAGS

@battre

**Comment by battre** [comment]
I cannot reproduce this. Can you give me a sequence of commands you execute after git clone? Also OS and compilers.

**Comment by oubayashi** [comment]
@gvollant

**Pull request merged**